src/v/cloud_storage/inventory/ntp_hashes.h

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "model/fundamental.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/iostream.hh>
+
+#include <absl/container/node_hash_set.h>
+
+namespace cloud_storage::inventory {
+
+enum class lookup_result {
+    exists,
+    missing,
+    possible_collision,
+};
+
+class ntp_path_hashes {
+public:
+    ntp_path_hashes(model::ntp ntp, std::filesystem::path hash_store_path);
+
+    ss::future<bool> load_hashes();
+
+    lookup_result exists(const remote_segment_path& path) const;
+
+private:
+    ss::future<> load_hashes(ss::input_stream<char>& stream);
+
+    model::ntp _ntp;
+    std::filesystem::path _hash_store_path;
+    std::filesystem::path _ntp_hashes_path;
+    absl::node_hash_set<uint64_t> _path_hashes;
```

Comment:
that could consume quite a lot of memory, why not btree_set?

Reply:
Since we do not need ordered semantics I used node hash set so insertion and queries are faster. I did not know that the node hash set uses more memory, we can use btree set instead if that uses less memory.

Reply:
The `node_hash_map` gives you pointer stability which means that it allocates a node for every element. The `flat_hash_map` would be ideal here but it will allocate contiguous memory. The `btree_set` uses b-tree to store keys and has small overhead per key (sizeof(key) * 1.4 IIRC). The lookup time is logarithmic though. Let's keep it this way and later we will be able to see if it consumes too much memory.

Reply:
actually, we have a `chunked_hash_set` in the tree now, it looks like it's using linear probing